### PR TITLE
[WIP] LLL: change alloc to use mload and fix alloc(0)

### PR DESCRIPTION
**Pull request by axic**
Depends on #2481.

### CODE REVIEW

**Comment by axic** [comment]
@benjaminion can you please check? (It depends on a couple of fixes first.)

The logic is: `mload(msize + n - 31) if (msize + n - 31) > 0`

You can work off this branch to if you want to play with it.

**Comment by benjaminion** [comment]
@axic, thanks, will do.

On this:

> LLL: use DUP2 instead of MSZIE to save gas

If I am reading the source and the YP correctly then MSIZE is 2 gas ("base" tier) and DUP2 is 3 gas ("very low" tier).

File: liblll/CodeFragment.cpp [hunk 1]
```diff
@@ -523,14 +523,24 @@ void CodeFragment::constructOperation(sp::utree const& _t, CompilerState& _s)
...
+			m_asm.append(Instruction::DUP3);
 			m_asm.append(Instruction::ADD);
+
+			/// Trigger memory expansion
 			m_asm.append(Instruction::SUB);
-			m_asm.append(Instruction::MSTORE8);
+			auto end = m_asm.newTag();
+			m_asm.append(u256(0));
+			m_asm.append(Instruction::DUP2);
+			m_asm.append(Instruction::LT);
```
Comment (benjaminion) [inline liblll/CodeFragment.cpp]: This needs be SLT rather than LT.

**Comment by axic** [comment]
@benjaminion correct, the tier naming is just very confusing, will remove that change :)

File: liblll/CodeFragment.cpp [hunk 2]
```diff
@@ -523,14 +523,24 @@ void CodeFragment::constructOperation(sp::utree const& _t, CompilerState& _s)
...
+			m_asm.appendJumpI(end);
+			m_asm.append(Instruction::MLOAD);
+			m_asm.append(end);
```
Comment (benjaminion) [inline liblll/CodeFragment.cpp]: We need to add a POP here after the end as both branches leave something on the stack unnecessarily (the output of the MLOAD or the result of the subtraction).

Even with these mods, the code allocs too much for (alloc 32) and not enough for N<31.  I'll have a go a reworking it.

**Comment by benjaminion** [comment]
The following passes all my test cases:

```
auto end = m_asm.newTag();
m_asm.append(Instruction::MSIZE); // Returns the original top of memory
m_asm.append(code[0].m_asm, 1);   // The alloc argument
m_asm.append(Instruction::DUP1);
m_asm.append(Instruction::ISZERO);// (alloc 0) is a no-op
m_asm.appendJumpI(end);
m_asm.append(u256(32));
m_asm.append(Instruction::DUP1);
m_asm.append(u256(1));
m_asm.append(Instruction::DUP4);  // The alloc argument
m_asm.append(Instruction::SUB);
m_asm.append(Instruction::DIV);
m_asm.append(Instruction::MUL);   // Align on 32 byte boundary
m_asm.append(Instruction::MSIZE); // MSIZE is cheap
m_asm.append(Instruction::ADD);
m_asm.append(Instruction::MLOAD);
m_asm.append(Instruction::POP);   // Discard the result of the MLOAD
m_asm.append(end);
m_asm.append(Instruction::POP);   // Discard the duplicate alloc arg
```

I don't think we can get away from treating `(alloc 0)` as a special case (if we're treating it as a no-op), and we need to be really careful when `(msize)` is initially 0. I think your approach won't be correct when msize=0 and we try to alloc less than 31 bytes.

**Comment by benjaminion** [comment]
This also works for slightly fewer opcodes, but much longer code output:

```
auto end = m_asm.newTag();
m_asm.append(Instruction::MSIZE); // Returns the original top of memory
m_asm.append(code[0].m_asm, 1);   // The alloc argument
m_asm.append(Instruction::DUP1);
m_asm.append(Instruction::ISZERO);// (alloc 0) is a no-op
m_asm.appendJumpI(end);
m_asm.append(u256(1));
m_asm.append(Instruction::DUP2);  // The alloc argument
m_asm.append(Instruction::SUB);
m_asm.append(u256(-32));
m_asm.append(Instruction::AND);   // Align on 32 byte boundary
m_asm.append(Instruction::MSIZE); // MSIZE is cheap
m_asm.append(Instruction::ADD);
m_asm.append(Instruction::MLOAD);
m_asm.append(Instruction::POP);   // Discard the result of the MLOAD
m_asm.append(end);
m_asm.append(Instruction::POP);   // Discard the duplicate alloc arg
```

**Comment by axic** [comment]
Replaced by #2545.

**Pull request closed without merging**